src/libraries/System.Private.Uri/src/System/UriExt.cs

```diff
@@ -552,28 +552,99 @@ internal unsafe bool InternalIsWellFormedOriginalString()
+        /// <returns>The unescaped representation of <paramref name="charsToUnescape"/>.</returns>
+        public static string UnescapeDataString(ReadOnlySpan<char> charsToUnescape)
+        {
+            return UnescapeDataString(charsToUnescape, backingString: null);
+        }
+
+        private static string UnescapeDataString(ReadOnlySpan<char> charsToUnescape, string? backingString = null)
+        {
+            Debug.Assert(backingString is null || backingString.Length == charsToUnescape.Length);
+
+            int indexOfFirstToUnescape = charsToUnescape.IndexOf('%');
+            if (indexOfFirstToUnescape < 0)
+            {
+                // Nothing to unescape, just return the original value.
+                return backingString ?? charsToUnescape.ToString();
+            }
+
+            var vsb = new ValueStringBuilder(stackalloc char[StackallocThreshold]);
+
+            // This may throw for very large inputs. ¯\_(ツ)_/¯
+            vsb.EnsureCapacity(charsToUnescape.Length - indexOfFirstToUnescape);
+
+            UriHelper.UnescapeString(
+                charsToUnescape.Slice(indexOfFirstToUnescape), ref vsb,
+                c_DummyChar, c_DummyChar, c_DummyChar,
+                UnescapeMode.Unescape | UnescapeMode.UnescapeAll,
+                syntax: null, isQuery: false);
+
+            string result = string.Concat(charsToUnescape.Slice(0, indexOfFirstToUnescape), vsb.AsSpan());
+            vsb.Dispose();
+            return result;
+        }
+
+        /// <summary>Attempts to convert a span to its unescaped representation.</summary>
+        /// <param name="charsToUnescape">The span to unescape.</param>
+        /// <param name="destination">The output span that contains the unescaped result of the operation.</param>
+        /// <param name="charsWritten">When this method returns, contains the number of chars that were written into <paramref name="destination"/>.</param>
+        /// <returns><see langword="true"/> if the <paramref name="destination"/> was large enough to hold the entire result; otherwise, <see langword="false"/>.</returns>
+        public static bool TryUnescapeDataString(ReadOnlySpan<char> charsToUnescape, Span<char> destination, out int charsWritten)
+        {
+            int indexOfFirstToUnescape = charsToUnescape.IndexOf('%');
+            if (indexOfFirstToUnescape < 0)
+            {
+                // Nothing to unescape, just copy the original chars.
+                if (charsToUnescape.TryCopyTo(destination))
+                {
+                    charsWritten = charsToUnescape.Length;
+                    return true;
+                }
 
-            int position = stringToUnescape.IndexOf('%');
-            if (position == -1)
-                return stringToUnescape;
+                charsWritten = 0;
+                return false;
+            }
 
             var vsb = new ValueStringBuilder(stackalloc char[StackallocThreshold]);
```

Comment:
Could we initialize the ValueStringBuilder with the destination buffer instead? Then at the end, if the VSB still contains the original span, we can avoid copying it, and if it doesn't, it presumably means the content was too big and we can fail?

Reply:
For the unescape path, most likely yes. I'll test how much of a difference it'd make.

For the escape path, we'd need a bit more code to handle overlapping input/output buffers. I'll measure it.

Reply:
Did it in 16f70f42a7fca46358acf4e92f24f415d4234a93, doesn't seem too bad.
It's slightly faster now.